include/sys/fdtable.h

```diff
@@ -113,14 +114,20 @@ void *z_get_fd_obj_and_vtable(int fd, const struct fd_op_vtable **vtable);
  * @param ... Variadic arguments to ioctl
  */
 static inline int z_fdtable_call_ioctl(const struct fd_op_vtable *vtable, void *obj,
-				       unsigned long request, ...)
+				       unsigned long request, int n_args, ...)
 {
-	va_list args;
-	int res;
+	va_list varargs;
+	int i, res;
+	uintptr_t args[3];
 
-	va_start(args, request);
-	res = vtable->ioctl(obj, request, args);
-	va_end(args);
+	__ASSERT_NO_MSG(n_args <= ARRAY_SIZE(args));
+
+	va_start(varargs, n_args);
```

Comment:
It's important here that the VA argument to ioctl can only ever be a single argument. Older specs took advantage of looser typing rules in C, and this is only really a way of allowing various pointer types to be passed to this function. It is quite reasonable to make this a single pointer argument.

It might also be helpful to look at the Linux implementation (or BSD), which uses various bits of the request to determine exactly how much data is expected in this argument, and even which direction data flows.

We definitely shouldn't be passing n_args/args through the systemcall. A single pointer is fine.

Reply:
> It might also be helpful to look at the Linux implementation (or BSD)

I don't try to "lift" implementation from other systems, together with associated copyright and licensing issues. I'm definitely aware of Linux maintainers' dislike for ioctl(), and understand why - just as ioctl() was a nice, clean, clever idea in the original Unix, just as it is unnice and unclear with all the "modern systems".

> We definitely shouldn't be passing n_args/args through the systemcall. A single pointer is fine.

Someone will need to elaborate in detail on those points.

It's otherwise pretty clear that syscall validation rules will need to be extended for the ioctl() case. That's why I'm questioning @andrewboie for details on both (intended) design and (actual) implementation of them.

Reply:
My point here is that that "..." in the ioctl call is a modern artifact, and that ioctl calls should always have exactly one pointer argument. The "..." is needed because it isn't really possible to get it right as some kind of generic pointer. But, there should never be an instance of 2, or more arguments.

Reply:
@d3zd3z:

> My point here is that that "..." in the ioctl call is a modern artifact

Can you elaborate on this a bit more? ("modern artifact")

> But, there should never be an instance of 2, or more arguments.

Regardless of the possible answer to the above question, the reason why ioctl() was brought in here is to exactly allow passing multiple arguments. ioctl() is used both for userspace->kernel and kernel->kernel communication. userspace->kernel subset if really lean so far - as you may imagine, nobody wants to add the complete Linux mess in there. The usecase behind supporting userspace ioctl() is, well, application portability POSIX-ish systems -> Zephyr. I.e. as people will port more apps, they will find ioctl() request types needed, and can implement then, then maybe even contribute upstream (maybe keep out of tree in forks). The ideas is, again, to provide a general interface which would work for any request type (and its params).

Now, kernel->kernel ioctl is used to generically implement some processing between implementations of the same interface (file descriptor) in kernel. E.g., there's a generic implementation of poll(), which uses kernel-ioctl to implement processing sub-steps polymorphically across different file descriptor objects. And ioctl is used there to save on the code size. The more ironically that protected userspace now adds much more overhead to it (i.e. no-split: code size savings comparing to "middle line", split kernel/userspace - code size bloat).

You're looking exactly at the support APIs for kernel->kernel ioctls.

All this may be not ideal, but nothing is ideal here, I look at support infra/docs for syscalls, and find it not ideal. This patch has a specific goal - set framework to allow all this work with kernel/userspace split. It doesn't do any further revolutions, where varargs calls were used before for kernel-kernel ioctl, there they're used now.

Reply:
> My point here is that that "..." in the ioctl call is a modern artifact, and that ioctl calls should always have exactly one pointer argument. The "..." is needed because it isn't really possible to get it right as some kind of generic pointer. But, there should never be an instance of 2, or more arguments.

@d3zd3z indeed, this is obvious from the documentation:

```
DESCRIPTION         top

       The ioctl() system call manipulates the underlying device parameters
       of special files.  In particular, many operating characteristics of
       character special files (e.g., terminals) may be controlled with
       ioctl() requests.  The argument fd must be an open file descriptor.

       The second argument is a device-dependent request code.  The third
       argument is an untyped pointer to memory.  It's traditionally char
       *argp (from the days before void * was valid C), and will be so named
       for this discussion.

       An ioctl() request has encoded in it whether the argument is an in
       parameter or out parameter, and the size of the argument argp in
       bytes.  Macros and defines used in specifying an ioctl() request are
       located in the file <sys/ioctl.h>.
```

ioctl() never been something you can send lots of arguments, and this is not ambiguous: 

> The third argument is an untyped pointer to memory.  It's traditionally char *argp (from the days before void * was valid C), and will be so named  for this discussion.

Reply:
> The third argument is an untyped pointer to memory.

Ok, as explained above, ->ioctl() is used to implement both userspace->kernel (let's call it just "ioctl") and kernel->kernel (let's call it "kioctl") functions. While userspace ioctl() accepts only one param, it's a matter of fact that kioctl() accepts of multiple. I explained why this was done - to save on the size of vtable pointers and function prologues/epilogues. Good or bad, that's how it is. If someone wants to change that, go ahead and submit a proposal and prototype refactor.

This patch is concerned with different matter - allowing ioctl() to work across userspace/kernel, while preserving how it works currently for kernel->kernel. (I agree that if userspace # of arg is fixed, the patch can be simplified. If there's something to say, then it's that I shouldn't have started on that while being assigned to other big and deep tasks, and multitasking among them, forgetting and overlooking such points.)